var/spack/repos/builtin/packages/arbor/package.py

```diff
@@ -13,26 +13,32 @@ class Arbor(CMakePackage, CudaPackage):
     homepage = "https://arbor-sim.org"
     git = "https://github.com/arbor-sim/arbor.git"
     url = "https://github.com/arbor-sim/arbor/releases/download/v0.8.1/arbor-v0.8.1-full.tar.gz"
-    maintainers("bcumming", "brenthuisman", "haampie", "schmitts")
+    maintainers = ["thorstenhater", "brenthuisman", "haampie"]
 
-    version("master", branch="master", submodules=True)
+    version("master", branch="master")
+    version("develop")
```

Comment:
There is no statement saying there is a "develop" branch?

There is a "develop" version, which is there to facilitate dev-builds.

Reply:
When I attempt `spack install arbor@develop` the fetchers fail:

```
==> Installing arbor-develop-m6das64hvefg5ovcjtoa4slpudtiwmk6 [33/33]
==> No binary for arbor-develop-m6das64hvefg5ovcjtoa4slpudtiwmk6 found: installing from source
==> Error: FetchError: All fetchers failed for spack-stage-arbor-develop-m6das64hvefg5ovcjtoa4slpudtiwmk6
```

Adding level 1 debugging, I get:

```
==> [2023-08-10-13:34:05.554038] Checking existence of https://mirror.spack.io/arbor/arbor-develop.tar.gz
==> [2023-08-10-13:34:05.802567] Failure reading URL: HEAD https://mirror.spack.io/arbor/arbor-develop.tar.gz returned 404: Not Found
==> [2023-08-10-13:34:05.802830] URL does not exist: https://mirror.spack.io/arbor/arbor-develop.tar.gz
==> [2023-08-10-13:34:05.803168] FailedDownloadError: Failed to fetch file from URL: https://mirror.spack.io/arbor/arbor-develop.tar.gz
==> [2023-08-10-13:34:05.803358] Checking existence of https://github.com/arbor-sim/arbor/releases/download/vdevelop/arbor-vdevelop-full.tar.gz
==> [2023-08-10-13:34:06.053692] Failure reading URL: HEAD https://github.com/arbor-sim/arbor/releases/download/vdevelop/arbor-vdevelop-full.tar.gz returned 404: Not Found
==> [2023-08-10-13:34:06.053888] URL does not exist: https://github.com/arbor-sim/arbor/releases/download/vdevelop/arbor-vdevelop-full.tar.gz
==> [2023-08-10-13:34:06.054074] FailedDownloadError: Failed to fetch file from URL: https://github.com/arbor-sim/arbor/releases/download/vdevelop/arbor-vdevelop-full.tar.gz
==> [2023-08-10-13:34:06.054146] Fetching from https://mirror.spack.io/arbor/arbor-develop.tar.gz failed.
==> [2023-08-10-13:34:06.054198] Fetching from https://github.com/arbor-sim/arbor/releases/download/vdevelop/arbor-vdevelop-full.tar.gz failed.
==> [2023-08-10-13:34:06.074428] Error: FetchError: All fetchers failed for spack-stage-arbor-develop-m6das64hvefg5ovcjtoa4slpudtiwmk6
Traceback (most recent call last):
  File "SPACK_ROOT/lib/spack/spack/build_environment.py", line 1050, in _setup_pkg_and_run
    return_value = function(pkg, kwargs)
  File "SPACK_ROOT/lib/spack/spack/installer.py", line 2501, in build_process
    return installer.run()
  File "SPACK_ROOT/lib/spack/spack/installer.py", line 2342, in run
    self.pkg.do_patch()
  File "SPACK_ROOT/lib/spack/spack/package_base.py", line 1452, in do_patch
    self.do_stage()
  File "SPACK_ROOT/lib/spack/spack/package_base.py", line 1437, in do_stage
    self.do_fetch(mirror_only)
  File "SPACK_ROOT/lib/spack/spack/package_base.py", line 1421, in do_fetch
    self.stage.fetch(mirror_only, err_msg=err_msg)
  File "SPACK_ROOT/lib/spack/spack/util/pattern.py", line 16, in __call__
    return [getattr(item, self.name)(*args, **kwargs) for item in self.container]
  File "SPACK_ROOT/lib/spack/spack/util/pattern.py", line 16, in <listcomp>
    return [getattr(item, self.name)(*args, **kwargs) for item in self.container]
  File "SPACK_ROOT/lib/spack/spack/stage.py", line 525, in fetch
    raise FetchError(err_msg or default_msg, None)
spack.util.web.FetchError: All fetchers failed for spack-stage-arbor-develop-m6das64hvefg5ovcjtoa4slpudtiwmk6
==> [2023-08-10-13:34:06.074985] Flagging arbor-develop-m6das64hvefg5ovcjtoa4slpudtiwmk6 as failed: FetchError: All fetchers failed for spack-stage-arbor-develop-m6das64hvefg5ovcjtoa4slpudtiwmk6
==> [2023-08-10-13:34:06.127413] ChildError: FetchError: All fetchers failed for spack-stage-arbor-develop-m6das64hvefg5ovcjtoa4slpudtiwmk6
```

Reply:
That's how I understood the `develop` tag is used: for local source builds (`dev-builds`). Since they don't have a version, they won't build otherwise, and you providing your own source is the point here.

Ideally, `develop` should be blocked when no local source is available. I can just hide the option after checking for source?

Reply:
> That's how I understood the `develop` tag is used: for local source builds (`dev-builds`). Since they don't have a version, they won't build otherwise, and you providing your own source is the point here.
> 
> Ideally, `develop` should be blocked when no local source is available. I can just hide the option after checking for source?

Do you mean the `develop` tag in the `spack.yaml` file for the developer workflow (https://spack-tutorial.readthedocs.io/en/latest/tutorial_developer_workflows.html.)?

There is more information on versions at https://spack.readthedocs.io/en/latest/packaging_guide.html#versions-and-fetching.  The directive you're using is specifically mentioned in the `Default Branch` subsection of https://spack.readthedocs.io/en/latest/packaging_guide.html#git.

Reply:
I meant the workflow for building from source provided not through Spack. As of [this change](https://github.com/spack/spack/pull/36273) in Spack, a Spack version needs to be provided for any build. We use the `dev-build` command, see https://github.com/arbor-sim/arbor/blob/master/.github/workflows/test-spack.yml for our workflow (we test PRs before merger in particular.

Reply:
Yep, that's how `spack dev-build` is documented at https://spack-tutorial.readthedocs.io/en/lanl19/tutorial_developer_workflows.html:
```ml
$ spack dev-build --until configure hdf5@develop +hl ~mpi
```